File: src/main/java/org/academy/internal/client/app/music/backend/decoder/mp3/ByteBufferInputStream.java
package org.academy.internal.client.app.music.backend.decoder.mp3;

import java.io.InputStream;
import java.nio.ByteBuffer;

public final class ByteBufferInputStream extends InputStream {
    private final ByteBuffer buffer;

    public ByteBufferInputStream(ByteBuffer buffer) {
        this.buffer = buffer;
    }

    @Override
    public int read() {
        if (!buffer.hasRemaining()) return -1;
        return buffer.get() & 0xFF;
    }

    @Override
    public int read(byte[] bytes, int off, int len) {
        if (!buffer.hasRemaining()) return -1;
        var length = Math.min(len, buffer.remaining());
        buffer.get(bytes, off, length);
        return length;
    }

    @Override
    public int available() {
        return buffer.remaining();
    }
}

File: src/main/kotlin/org/academy/internal/client/app/music/decoder/ByteBufferInputStream.kt
package org.academy.internal.client.app.music.decoder

import java.io.InputStream
import java.nio.ByteBuffer
import kotlin.math.min

class ByteBufferInputStream(private val buffer: ByteBuffer) : InputStream() {
    override fun read(): Int {
        if (!buffer.hasRemaining()) return -1
        return buffer.get().toInt() and 0xFF
    }

    override fun read(bytes: ByteArray, off: Int, len: Int): Int {
        if (!buffer.hasRemaining()) return -1
        val length = min(len, buffer.remaining())
        buffer.get(bytes, off, length)
        return length
    }

    override fun available(): Int {
        return buffer.remaining()
    }
}